Sources/Core/Tracker/WebViewMessageHandlerV2.swift

```diff
@@ -0,0 +1,142 @@
+                pingYOffsetMin: atomicJson["pingYOffsetMin"] as? Int ?? nil,
+                pingYOffsetMax: atomicJson["pingYOffsetMax"] as? Int ?? nil
+            )
+            
+            track(event, withEntities: entitiesJson, andTrackers: trackers)
+        }
+    }
+    
+    func track(_ event: Event, withEntities entities: [[AnyHashable : Any]], andTrackers trackers: [String]) {
+        event.entities = parseEntities(entities)
+        
+        if trackers.count > 0 {
+            for namespace in trackers {
+                if let tracker = Snowplow.tracker(namespace: namespace) {
+                    _ = tracker.track(event)
+                }
+            }
+        } else {
+            _ = Snowplow.defaultTracker()?.track(event)
+        }
+    }
+    
+    func createSelfDescribingJson(_ map: [AnyHashable : Any]) -> SelfDescribingJson? {
+        if let schema = map["schema"] as? String,
+           let payload = map["data"] as? [String : Any] {
+            return SelfDescribingJson(schema: schema, andDictionary: payload)
+        }
+        return nil
+    }
+    
+    func parseEntities(_ entities: [[AnyHashable : Any]]) -> [SelfDescribingJson] {
+        var contextEntities: [SelfDescribingJson] = []
+        
+        for entityJson in entities {
+            if let entity = createSelfDescribingJson(entityJson) {
+                contextEntities.append(entity)
+            }
+        }
+        return contextEntities
+    }
+    
+    func parseAtomicPropertiesFromMessage(_ messageString: String?) -> [String : Any]? {
+        guard let atomicData = messageString?.data(using: .utf8) else {
+            logError(message: "WebView: No atomic properties provided, skipping.")
+            return nil
+        }
+        guard let atomicJson = try? JSONSerialization.jsonObject(with: atomicData) as? [String : Any] else {
+            logError(message: "WebView: Received event payload is not serializable to JSON, skipping.")
+            return nil
+        }
+        return atomicJson
+    }
+    
+    func parseSelfDescribingEventDataFromMessage(_ messageString: String?) -> [String : Any]? {
+        guard (messageString != nil) else { return nil }
+        guard let eventData = messageString?.data(using: .utf8) else {
+            logError(message: "WebView: Received event payload is not serializable to JSON, skipping.")
+            return nil
+        }
+        guard let eventJson = try? JSONSerialization.jsonObject(with: eventData) as? [String : Any] else {
+            logError(message: "WebView: Received event payload is not serializable to JSON, skipping.")
+            return nil
+        }
+        return eventJson
+    }
+    
+    func parseEntitiesFromMessage(_ messageString: String?) -> [[AnyHashable : Any]]? {
+        guard (messageString != nil) else { return nil }
+        guard let entitiesData = messageString?.data(using: .utf8) else {
```

Comment:
You can chain these to avoid code duplication:

```swift
        guard let entitiesData = messageString?.data(using: .utf8),
              let entitiesJson = try? JSONSerialization.jsonObject(with: entitiesData) as? [[AnyHashable : Any]] else {
            logError(message: "WebView: Received event payload is not serializable to JSON, skipping.")
            return nil
        }
```